src/aui/auibar.cpp

```diff
@@ -2207,60 +2207,67 @@ void wxAuiToolBar::DoIdleUpdate()
         if (item.m_toolId == -1)
             continue;
 
-        wxUpdateUIEvent evt(item.m_toolId);
-        evt.SetEventObject(this);
+        if (item.GetKind() != wxITEM_CONTROL)
```

Comment:
Do the checks for `item.m_window` inside this check still make sense?

I don't quite follow all the ramifications of this change, e.g. is the window ID always guaranteed to be the same as `item.m_toolId` in this case? If not, we might be sending different events now.

Perhaps it would be safer to add, and call, `item.Is3State()` here instead?

Reply:
1)  The code for `wxAuiToolBar::AddControl()` implies to me that `item.m_toolId` ought to always be equal to `wxControl::GetId()`
2)  The code for `wxAuiToolBar::AddControl()` and `wxAuiToolBar::AddTool()` imply to me that `item.m_kind` ought to be `wx_ITEM_CONTROL` if and only if `AddControl()` (not `AddTool()`) was used, and also that `item.m_window` ought to be `!= nullptr` if and only if `AddControlI()` was used
3)  Unfortunately, the `wxAuitToolBarItem::SetId()`, `wxAuitToolBarItem::SetKind()`, and `wxAuitToolBarItem::SetWindow()` functions allow these ought-to-be-invariants to be violated.
4)  The unmodified code only respects `wxUpdateUIEvent::Check()` for `wxITEM_CHECK` and `wxITEM_RADIO`.  In other words, the unmodified code doesn't properly handle `wxCHK_2STATE wxCheckBox`es because they are `wxITEM_CONTROL`.
5)  The unmodified code doesn't support `wxUpdateUIEvent::Show()` or `wxUpdateUIEvent::SetText()` for `wxITEM_CONTROL` even though it seems to me that those operations should be supported.

I think the code is currently not in a very good state, but I don't see a way to completely fix it without the breaking change of eliminating `wxAuitToolBarItem::SetId()`, `wxAuitToolBarItem::SetKind()`, and `wxAuitToolBarItem::SetWindow()`.  My opinion is to either entirely ignore those functions (which is what I did since I was only trying to make `wxCheckBox` work), or else add `wxCHECK()`s that would treat changing from a `wxITEM_CONTROL` to a non-`wxITEM_CONTROL` (or vice versa) an error, and also that would enforce `item.m_window != nullptr` if and only `item.m_kind == wxITEM_CONTROL`.

I prefer that all `wxControl`s, not just `wxCHK_3STATE wxCheckBox`es, use `wxWindow::UpdateWindowUI()`, but I don't get to make the final decision.  If this message still hasn't convinced you, I will work through the `item.Is3State()` concept.  On the other hand, if I have convinced you, then I should, as you implied, remove the `item.m_window` checks.

Reply:
I also hadn't realized the control `wxUpdateUIEvent` handling was so restricted until after I implemented `wxUpdateUIEvent::Set3StateValue()` and realized it was having any effect.

I don't actually need the additional `wxUpdateUIEvent` functionality for other controls, but knowing these limitations exist is irritating me, so I would prefer a general solution.  Also, it seems weird to me to make a change adding `item.Is3State()` just to submit another PR to remove it again when making the general replacement.  If you would prefer the general `wxITEM_CONTROL` changes as a separate PR from this one (which is reasonable), I am actually inclined to submit that change BEFORE this one so that we don't wind up with this very briefly existing special case for `item.Is3State()`.  Is that OK with you?

Reply:
Sure, 2 PRs would be fine, but please also feel free to submit it even as part of the same PR but as a separate commit _preceding_ the rest of the changes — I think this should make things tidier, and this commit might be backported to 3.2 too, if desired.

Thanks again!

Reply:
> Sure, 2 PRs would be fine, but please also feel free to submit it even as part of the same PR but as a separate commit _preceding_ the rest of the changes — I think this should make things tidier, and this commit might be backported to 3.2 too, if desired.

OK, I will re-sequence the commits in this PR.

Reply:
changes force-pushed